InfrastructureBackupValidator/BackupValidationTool.psm1

```diff
@@ -0,0 +1,591 @@
+
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [String]
+        $SQLServerInstanceName = "localhost\SQLEXPRESS", # Use default SQL Express server instance
+
+        [Parameter(Mandatory = $false)]
+        [ValidateNotNullOrEmpty()]
+        [PSCredential]
+        $SQLCredential,
+
+        [Parameter(Mandatory = $false)]
+        [ValidateScript({$_ | Test-Path -PathType Container})]
+        [String]
+        $TempFolder = $env:TEMP
+    )
+
+    $ErrorActionPreference = "Stop"
+
+    # STEP 1: Connect to the backup store with provided credential
+    Write-Verbose $Strings.ProgressConnectBackupStore -Verbose
+    try
+    {
+        $null = New-PSDrive -Name $BackupStoreDriveName -Root $BackupStorePath -Credential $BackupStoreCredential -PSProvider FileSystem | Out-Null
+    }
+    catch
+    {
+        throw ($Strings.ErrorFailToConnectBackupStore -f $BackupStorePath, $_)
+    }
+
+    Install-Module -Name SqlServer
+    Import-Module SqlServer -DisableNameChecking
+    $sqlCommonParams = @{
+        ServerInstance = $SQLServerInstanceName
+    }
+
+    if (!$SQLCredential)
+    {
+        # STEP 2: Connect to the SQL server with Windows Authentication
+        Write-Verbose ($Strings.ProgressConnectSqlServerDefault -f $SQLServerInstanceName) -Verbose
+        try
+        {
+            $sqlInstance = Get-SqlInstance -ServerInstance $SQLServerInstanceName
+        }
+        catch
+        {
+            throw ($Strings.ErrorFailToConnectSqlServerDefault -f $SQLServerInstanceName, $_)
+        }
+    }
+    else
+    {
+        # STEP 2: Connect to the SQL server with provided credential
+        Write-Verbose ($Strings.ProgressConnectSqlServerProvided -f $SQLServerInstanceName) -Verbose
+        try
+        {
+            $sqlInstance = Get-SqlInstance -ServerInstance $SQLServerInstanceName -Credential $SQLCredential
+            $sqlCommonParams.SQLCredential = $SQLCredential
+        }
+        catch
+        {
+            throw ($Strings.ErrorFailToConnectSqlServerProvided -f $SQLServerInstanceName, $_)
+        }
+    }
+
+    # STEP 3: Create a tmp folder
+    Write-Verbose ($Strings.ProgressCreateTmpFolder -f $SubTmpFolderName, $TempFolder) -Verbose
+    $tmpDir = Join-Path $TempFolder $SubTmpFolderName
+    $null = Remove-Item $tmpDir -Force -Recurse -ErrorAction Ignore | Out-Null
+    $null = New-Item -Type Directory $tmpDir -Force | Out-Null
+    
+    # STEP 4: Get the list of backup files according to the BackupID
+    Write-Verbose ($Strings.ProgressGetBackupSnapshots -f $BackupID) -Verbose
+    $backupRoot = Join-Path $BackupStorePath "MASBackup\progressivebackup"
+    $backupFiles = (Get-ChildItem -Path $backupRoot -Recurse | ? { $_.Name -match $BackupID -and $_.Name -match ".zip" }).FullName
+    $backupRepos = @($BackupRepoNames.Values.GetEnumerator()) | ? { $_ -ne $BackupRepoNames.Subscription }
+    $backupFiles = $backupFiles | ? { $backupRepos.Contains(($_ -Split "\\")[-2]) }
```

Comment:
Could you add an example repo filename to help readers understand better which element is getting extracted from the file path?

Reply:
Will add this to the code comment:
A backup file FullName is like: "$BackupStorePath\MASBackup\progressivebackup\1.2209.0.53\NRP;Microsoft.Network.Admin;-;Quota\202301171931_fa396b1d-1814-4c75-9ddc-fef71cf2ecd1_Full_C.zip"
($_ -Split "\\")[-1] would be "202301171931_fa396b1d-1814-4c75-9ddc-fef71cf2ecd1_Full_C.zip" containing the backup ID.
($_ -Split "\\")[-2] would be "NRP;Microsoft.Network.Admin;-;Quota" which is the name of the repo.
($_ -Split "\\")[-3] would be "1.2209.0.53" which is the name of the backup store, also the stamp version.